doxygen/man/man3/Reel.3
.TH "Reel" 3 "Samedi Juin 16 2012" "Calculatrice polonais inversé LO21" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Reel \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <reel\&.h>\fP
.PP
Est dérivée de \fBConstante\fP\&.
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "\fBReel\fP (double e)"
.br
.ti -1c
.RI "\fBReel\fP (\fBConstante\fP *c)"
.br
.ti -1c
.RI "virtual \fB~Reel\fP ()"
.br
.ti -1c
.RI "virtual QString \fBafficher\fP () const "
.br
.ti -1c
.RI "double \fBgetValeur\fP () const "
.br
.ti -1c
.RI "void \fBsetValeur\fP (int v)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBaddition\fP (\fBConstante\fP *c)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBproduit\fP (\fBConstante\fP *c)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBdivision\fP (\fBConstante\fP *c)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBsigne\fP ()"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBsoustraction\fP (\fBConstante\fP *c)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBinv\fP ()"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBfact\fP ()"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBsinus\fP (bool angle)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBcosinus\fP (bool angle)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBsinush\fP (bool angle)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBcosinush\fP (bool angle)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBtangente\fP (bool angle)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBtangenteh\fP (bool angle)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBlogN\fP ()"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBlog1\fP ()"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBpuissance\fP (\fBConstante\fP *c)"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBcarre\fP ()"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBcube\fP ()"
.br
.ti -1c
.RI "virtual \fBConstante\fP * \fBracine\fP ()"
.br
.in -1c
.SH "Description détaillée"
.PP 
Définition à la ligne 7 du fichier reel\&.h\&.
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "\fBReel::Reel\fP (doublee)\fC [inline]\fP"
.PP
Définition à la ligne 12 du fichier reel\&.h\&.
.SS "\fBReel::Reel\fP (\fBConstante\fP *c)"Constructeur de réel
.PP
On vérifie le type de la constante passée en paramètre, et si c'est un réel, on recopie sa valeur dans le nouvel entier, sinon on lève une exception\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIc\fP Une constante, donc entier, réel, rationnel ou complexe
.RE
.PP

.PP
Définition à la ligne 21 du fichier reel\&.cpp\&.
.SS "virtual \fBReel::~Reel\fP ()\fC [inline, virtual]\fP"
.PP
Définition à la ligne 14 du fichier reel\&.h\&.
.SH "Documentation des fonctions membres"
.PP 
.SS "\fBConstante\fP * \fBReel::addition\fP (\fBConstante\fP *c)\fC [virtual]\fP"Somme de 2 réels
.PP
On vérifie le type du paramètre, et on crée une nouvelle instance de ce type en fonction, dont le ou les attributs seront en fait la somme des 2 réels dont l'on veut calculer l'addition\&.Cependant, si on est en mode Réel, on ne doit pas avoir les autres types de constante\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIc\fP Une constante 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit, car on utilisera toujours les formes complexes pour les calculs
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 52 du fichier reel\&.cpp\&.
.SS "virtual QString \fBReel::afficher\fP () const\fC [inline, virtual]\fP"
.PP
Implémente \fBConstante\fP\&.
.PP
Définition à la ligne 15 du fichier reel\&.h\&.
.SS "\fBConstante\fP * \fBReel::carre\fP ()\fC [virtual]\fP"Fonction carré
.PP
On crée un nouveau Réel dont la valeur est le résultat du carré 
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit à partir du résultat de la fonction
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 287 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::cosinus\fP (boolangle)\fC [virtual]\fP"Calcul du Cosinus
.PP
Si le paramètre est vrai, on se situe en mode Degré\&. Dans ce cas, il faut multiplier la valeur par PI/180, car la fonction \fIcos\fP de cmath est définie en radians\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIangle\fP Sert à savoir si l'on est en mode Degré ou non 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit construit à partir d'un entier contenant la valeur du cosinus
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 161 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::cosinush\fP (boolangle)\fC [virtual]\fP"Calcul du Cosinus Hyperbolique
.PP
Si le paramètre est vrai, on se situe en mode Degré\&. Dans ce cas, il faut multiplier la valeur par PI/180, car la fonction \fIcosh\fP de cmath est définie en radians\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIangle\fP Sert à savoir si l'on est en mode Degré ou non 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit construit à partir d'un entier contenant la valeur du cosinus hyperbolique
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 176 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::cube\fP ()\fC [virtual]\fP"Fonction cube
.PP
On crée un nouveau Réel dont la valeur est le résultat du cube 
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit à partir du résultat de la fonction
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 298 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::division\fP (\fBConstante\fP *c)\fC [virtual]\fP"Division de réels
.PP
Comme pour la somme et le produit, on vérifie le type du paramètre, et on crée une nouvelle instance de ce type qui contiendra le quotient demandé\&. Cependant, normalement en mode Réel, on ne doit avoir que des réels\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIc\fP Une \fBConstante\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit contenant le quotient
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 104 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::fact\fP ()\fC [virtual]\fP"
.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 236 du fichier reel\&.cpp\&.
.SS "double \fBReel::getValeur\fP () const\fC [inline]\fP"
.PP
Définition à la ligne 16 du fichier reel\&.h\&.
.SS "\fBConstante\fP * \fBReel::inv\fP ()\fC [virtual]\fP"Inversion d'un réel
.PP
Pour un réel, on peut créer un nouveau réel contenant l'inverse de celui que l'on veut 
.PP
\fBRenvoie:\fP
.RS 4
Un \fIRéel\fP contenant le résultat
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 136 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::log1\fP ()\fC [virtual]\fP"Logarithme décimal
.PP
On crée un nouveau Réel dont la valeur est celle du logarithme décimal du réel initial\&. 
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit avec ce réel
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 254 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::logN\fP ()\fC [virtual]\fP"Logarithme népérien
.PP
On crée un nouveau Réel dont la valeur est celle du logarithme népérien du réel initial\&. 
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit avec ce réel
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 243 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::produit\fP (\fBConstante\fP *c)\fC [virtual]\fP"Produit de deux réels
.PP
Comme pour la somme, on vérifie le type du paramètre, et on crée une nouvelle instance de ce type qui contiendra le produit demandé\&. Cependant, normalement en mode Réel, on ne doit avoir que des réels\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIc\fP Une \fBConstante\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit contenant le produit
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 77 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::puissance\fP (\fBConstante\fP *c)\fC [virtual]\fP"Puissance de deux réels
.PP
On effectue un transtypage en réel afin de récupérer l'exposant sous forme réelle\&. On calcule ensuite grâce à la fonction \fIpow\fP de la bibliothèque \fIcmath\fP\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIc\fP Une \fI\fBConstante\fP\fP qui sera l'exposant 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit à partir du résultat
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 265 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::racine\fP ()\fC [virtual]\fP"Fonction racine carrée
.PP
On crée un nouveau Réel dont la valeur est le résultat de la racine carrée 
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit à partir du résultat de la fonction
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 309 du fichier reel\&.cpp\&.
.SS "void \fBReel::setValeur\fP (intv)\fC [inline]\fP"
.PP
Définition à la ligne 17 du fichier reel\&.h\&.
.SS "\fBConstante\fP * \fBReel::signe\fP ()\fC [virtual]\fP"Inversion de signe d'un réel
.PP
On crée un nouvel entier dont la valeur est l'opposée de celle initiale 
.PP
\fBRenvoie:\fP
.RS 4
Un \fIRéel\fP, de valeur opposée
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 116 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::sinus\fP (boolangle)\fC [virtual]\fP"Calcul du Sinus
.PP
Si le paramètre est vrai, on se situe en mode Degré\&. Dans ce cas, il faut multiplier la valeur par PI/180, car la fonction \fIsin\fP de cmath est définie en radians\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIangle\fP Sert à savoir si l'on est en mode Degré ou non 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit construit à partir d'un entier contenant la valeur du sinus
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 146 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::sinush\fP (boolangle)\fC [virtual]\fP"Sinus Hyperbolique Réel
.PP
Si le paramètre est vrai, on se situe en mode Degré\&. Dans ce cas, il faut multiplier la valeur par PI/180, car la fonction \fItanh\fP de cmath est définie en radians\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIangle\fP Booléen, permet de savoir si l'on est en mode degré 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un nouveau \fI\fBReel\fP\fP construit à partir du \fBReel\fP contenant le résultat
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 222 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::soustraction\fP (\fBConstante\fP *c)\fC [virtual]\fP"Différence de réels
.PP
Comme pour les autres opérations, on vérifie le type du paramètre, et on crée une nouvelle instance de ce type qui contiendra la différence voulue\&. Cependant, normalement en mode Réel, on ne doit avoir que des réels\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIc\fP Une \fBConstante\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit contenant la différence
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 125 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::tangente\fP (boolangle)\fC [virtual]\fP"Calcul de la Tangente
.PP
Si le paramètre est vrai, on se situe en mode Degré\&. Dans ce cas, il faut multiplier la valeur par PI/180, car la fonction \fItan\fP de cmath est définie en radians\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIangle\fP Sert à savoir si l'on est en mode Degré ou non 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP construit construit à partir d'un entier contenant la valeur de la tangente
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 192 du fichier reel\&.cpp\&.
.SS "\fBConstante\fP * \fBReel::tangenteh\fP (boolangle)\fC [virtual]\fP"Calcul de la Tangente Hyperbolique
.PP
Si le paramètre est vrai, on se situe en mode Degré\&. Dans ce cas, il faut multiplier la valeur par PI/180, car la fonction \fItanh\fP de cmath est définie en radians\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIangle\fP Sert à savoir si l'on est en mode Degré ou non 
.RE
.PP
\fBRenvoie:\fP
.RS 4
Un \fI\fBReel\fP\fP la valeur de la tangente hyperbolique
.RE
.PP

.PP
Réimplémentée à partir de \fBConstante\fP\&.
.PP
Définition à la ligne 207 du fichier reel\&.cpp\&.

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour Calculatrice polonais inversé LO21 à partir du code source\&.


reel.cpp
#include "complexe.h"
#include "entier.h"
#include "rationnel.h"
#include "reel.h"
#include <typeinfo>
#include <QString>
#include <iostream>
#define PI 3.14159265

/*
Rationnel Reel::toRationnel() const {
    //DANS LE BUT DE NE PAS PERDRE DE PRÉCISION LORS DES DIVISIONS ON A BESOIN DE TRANSFORMER LES RÉELS EN RATIONNELS
    int partie_entiere = floor(this->getValeur());
    float partie_decimale = this->getValeur()-partie_entiere;
    std::cout<<partie_decimale;

    //NUL => ETANT DONNE LA MEMORISATION DU FLOAT (virgule flottante) CA SERVIRA A RIEN
}
*/

Reel::Reel(Constante* c) {
    if (Entier *c_entier=dynamic_cast<Entier *>(c)){
        _valeur = c_entier->getValeur();
    } else if (typeid(*c)==typeid(Complexe)) {
        Complexe *c_complexe=dynamic_cast<Complexe *>(c);
        if (c_complexe->reel_pur()){
            if (Entier *re_entiere=dynamic_cast<Entier *>(c_complexe->getPartieReelle())){
                _valeur = re_entiere->getValeur();
            } else if (Reel *re_reelle=dynamic_cast<Reel *>(c_complexe->getPartieReelle())){
                _valeur = re_reelle->getValeur();
            } else {
                throw LogMessage(3,"Seuls le complexes dont la partie réelle est entière ou réelle et la partie imaginaire est nulle permettent de construire un réel.", important);
            }
        } else {
            throw LogMessage(3,"Seuls le complexes dont la partie réelle est entière ou réelle et la partie imaginaire est nulle permettent de construire un réel.", important);
        }
        throw LogMessage(3,"Impossible de construir un réel à partir d'un complexe.", important);
    } else if (typeid(*c)==typeid(Rationnel)) {
        throw LogMessage(3,"Impossible de construir un réel à partir d'un rationnel sans perte de précision.", important);
    } else if (typeid(*c)==typeid(Reel)) {
        Reel *c_reel=dynamic_cast<Reel *>(c);
        _valeur = c_reel->getValeur();
    }
    /**
      * \brief Constructeur de réel
      * \details On vérifie le type de la constante passée en paramètre, et si c'est un réel, on
      *          recopie sa valeur dans le nouvel entier, sinon on lève une exception.
      * \param c Une constante, donc entier, réel, rationnel ou complexe
      */
}

Constante* Reel::addition(Constante* c){
    if (Entier *c_entier=dynamic_cast<Entier *>(c)){
        return c_entier->addition(this);
    } else if (typeid(*c)==typeid(Complexe)) {
        Complexe *c_complexe=dynamic_cast<Complexe *>(c);
        return c_complexe->addition(this);
    } else if (typeid(*c)==typeid(Rationnel)) {
        Rationnel *c_rationnel=dynamic_cast<Rationnel *>(c);
        throw LogMessage(3,"Impossible d'additionner un réel et un rationnel sans perte de précision.", important);
        return c_rationnel->addition(this);
    } else if (typeid(*c)==typeid(Reel)) {
        Reel *c_reel=dynamic_cast<Reel *>(c);
        return new Reel(c_reel->getValeur()+_valeur);
    }
    throw LogMessage(3,"Type de constante non prévu dans la fonction Constante* Reel::addition(Constante *c).", important);
    /**
      * \brief Somme de 2 réels
      * \details On vérifie le type du paramètre, et on crée une nouvelle instance de ce type en fonction, dont le
      *          ou les attributs seront en fait la somme des 2 réels dont l'on veut calculer l'addition.Cependant, si on
      *          est en mode Réel, on ne doit pas avoir les autres types de constante.
      * \param c Une constante
      * \return Un \e Reel construit, car on utilisera toujours les formes complexes pour les calculs
      */
}

Constante* Reel::produit(Constante *c)
{
    if (Entier *c_entier=dynamic_cast<Entier *>(c)){
        return c_entier->produit(this);
    } else if (typeid(*c)==typeid(Rationnel)) {
        Rationnel *c_rationnel=dynamic_cast<Rationnel *>(c);
        return c_rationnel->produit(this);
    }
    else if (typeid(*c)==typeid(Reel)) {
            Reel *c_reel=dynamic_cast<Reel *>(c);
            return new Reel(_valeur*c_reel->getValeur());
        }
    else if (typeid(*c)==typeid(Complexe)) {
             Complexe *c_complexe=dynamic_cast<Complexe *>(c);
             return c_complexe->produit(this);
     }
    throw LogMessage(3,"Type de constante non prévu dans la fonction Constante* Reel::produit(Constante *c).", important);
    /**
      * \brief Produit de deux réels
      * \details Comme pour la somme, on vérifie le type du paramètre, et on crée une nouvelle instance de ce type qui contiendra
      *          le produit demandé. Cependant, normalement en mode Réel, on ne doit avoir que des réels.
      * \param c Une Constante
      * \return Un \e Reel construit contenant le produit
      */
}


Constante* Reel::division(Constante *c)
{
    return this->produit(c->inv());
   /**
     * \brief Division de réels
     * \details Comme pour la somme et le produit, on vérifie le type du paramètre, et on crée une nouvelle instance de ce type qui contiendra
     *          le quotient demandé. Cependant, normalement en mode Réel, on ne doit avoir que des réels.
     * \param c Une Constante
     * \return Un \e Reel construit contenant le quotient
     */
}

Constante* Reel::signe(){
    return new Reel(-_valeur);
    /**
      * \brief Inversion de signe d'un réel
      * \details On crée un nouvel entier dont la valeur est l'opposée de celle initiale
      * \return Un \e Réel, de valeur opposée
      */
}

Constante* Reel::soustraction(Constante* c){
    return this->addition(c->signe());
    /**
      * \brief Différence de réels
      * \details Comme pour les autres opérations, on vérifie le type du paramètre, et on crée une nouvelle instance de ce type qui contiendra
      *          la différence voulue. Cependant, normalement en mode Réel, on ne doit avoir que des réels.
      * \param c Une Constante
      * \return Un \e Reel construit contenant la différence
      */
}

Constante *Reel::inv()
{
    if (this->getValeur()==0) {
          throw LogMessage(1,"Division par zéro", important);
          return this;
    }
    return new Reel(1/this->getValeur());
    /**
      * \brief Inversion d'un réel
      * \details Pour un réel, on peut créer un nouveau réel contenant l'inverse de celui que l'on veut
      * \return Un \e Réel contenant le résultat
      */
}

Constante* Reel::sinus(bool angle)
{
    float res = _valeur;
    if(angle)
        res=res*PI/180;
    return new Reel(sin(res));
    /**
      * \brief Calcul du Sinus
      * \details Si le paramètre est vrai, on se situe en mode Degré. Dans ce cas, il faut multiplier la valeur par PI/180,
      *          car la fonction \e sin de cmath est définie en radians.
      * \param angle Sert à savoir si l'on est en mode Degré ou non
      * \return Un \e Reel construit construit à partir d'un entier contenant la valeur du sinus
      */
}

Constante* Reel::cosinus(bool angle)
{
    float res = _valeur;
    if(angle)
        res=res*PI/180;
    return new Reel(cos(res));
    /**
      * \brief Calcul du Cosinus
      * \details Si le paramètre est vrai, on se situe en mode Degré. Dans ce cas, il faut multiplier la valeur par PI/180,
      *          car la fonction \e cos de cmath est définie en radians.
      * \param angle Sert à savoir si l'on est en mode Degré ou non
      * \return Un \e Reel construit construit à partir d'un entier contenant la valeur du cosinus
      */
}

Constante* Reel::cosinush(bool angle)
{
    float res = _valeur;
    if(angle)
        res=res*PI/180;
    return new Reel(cosh(res));
    /**
      * \brief Calcul du Cosinus Hyperbolique
      * \details Si le paramètre est vrai, on se situe en mode Degré. Dans ce cas, il faut multiplier la valeur par PI/180,
      *          car la fonction \e cosh de cmath est définie en radians.
      * \param angle Sert à savoir si l'on est en mode Degré ou non
      * \return Un \e Reel construit construit à partir d'un entier contenant la valeur du cosinus hyperbolique
      */
}


Constante* Reel::tangente(bool angle)
{
    float res = _valeur;
    if(angle)
        res=res*PI/180;
    return new Reel(tan(res));
    /**
      * \brief Calcul de la Tangente
      * \details Si le paramètre est vrai, on se situe en mode Degré. Dans ce cas, il faut multiplier la valeur par PI/180,
      *          car la fonction \e tan de cmath est définie en radians.
      * \param angle Sert à savoir si l'on est en mode Degré ou non
      * \return Un \e Reel construit construit à partir d'un entier contenant la valeur de la tangente
      */
}

Constante* Reel::tangenteh(bool angle)
{
    float res = _valeur;
    if(angle)
        res=res*PI/180;
    return new Reel(tanh(res));
    /**
      * \brief Calcul de la Tangente Hyperbolique
      * \details Si le paramètre est vrai, on se situe en mode Degré. Dans ce cas, il faut multiplier la valeur par PI/180,
      *          car la fonction \e tanh de cmath est définie en radians.
      * \param angle Sert à savoir si l'on est en mode Degré ou non
      * \return Un \e Reel la valeur de la tangente hyperbolique
      */
}

Constante* Reel::sinush(bool angle)
{
    float res = _valeur;
    if(angle)
        res=res*PI/180;
    return new Reel(sinh(res));
    /**
      * \brief Sinus Hyperbolique Réel
      * \details Si le paramètre est vrai, on se situe en mode Degré. Dans ce cas, il faut multiplier la valeur par PI/180,
      *          car la fonction \e tanh de cmath est définie en radians.
      * \param angle Booléen, permet de savoir si l'on est en mode degré
      * \return Un nouveau \e Reel construit à partir du Reel contenant le résultat
      */
}
Constante* Reel::fact()
{
    throw LogMessage(2,"La fonction factorielle n'est pas implémentée pour les réels.", moyen);
    return this;
}


Constante* Reel::logN()
{
    float res = log(_valeur);
    return new Reel(res);
    /**
      * \brief Logarithme népérien
      * \details On crée un nouveau Réel dont la valeur est celle du logarithme népérien du réel initial.
      * \return Un \e Reel construit avec ce réel
      */
}

Constante* Reel::log1()
{
    float res = log10(_valeur);
    return new Reel(res);
    /**
      * \brief Logarithme décimal
      * \details On crée un nouveau Réel dont la valeur est celle du logarithme décimal du réel initial.
      * \return Un \e Reel construit avec ce réel
      */
}

Constante* Reel::puissance(Constante *c)
{
    if (typeid(*c)==typeid(Reel)) {
            //transtypage en entier
            Reel *c_reel=dynamic_cast<Reel *>(c);
            int value;
            value = c_reel->getValeur();
    float res;
    float res2 = _valeur;
    res = pow(res2,value);

    return new Reel(res);
    }
    /**
      * \brief Puissance de deux réels
      * \details On effectue un transtypage en réel afin de récupérer l'exposant sous forme réelle.
      *          On calcule ensuite grâce à la fonction \e pow de la bibliothèque \e cmath.
      * \param c Une \e Constante qui sera l'exposant
      * \return Un \e Reel construit à partir du résultat
      */
}

Constante* Reel::carre()
{
    float res = _valeur*_valeur;
    return new Reel(res);
    /**
      * \brief Fonction carré
      * \details On crée un nouveau Réel dont la valeur est le résultat du carré
      * \return Un \e Reel construit à partir du résultat de la fonction
      */
}

Constante* Reel::cube()
{
    float res = _valeur*_valeur*_valeur;
    return new Reel(res);
    /**
      * \brief Fonction cube
      * \details On crée un nouveau Réel dont la valeur est le résultat du cube
      * \return Un \e Reel construit à partir du résultat de la fonction
      */
}

Constante* Reel::racine()
{
    float res = sqrt(_valeur);
    return new Reel(res);
    /**
      * \brief Fonction racine carrée
      * \details On crée un nouveau Réel dont la valeur est le résultat de la racine carrée
      * \return Un \e Reel construit à partir du résultat de la fonction
      */
}

